### PR TITLE
Alert Feature for Discord Bot Bounty

### PR DESCRIPTION
**Pull request by ghost**
Added features to allow for people to set up alerts in channels in their guilds for their set default coins.
Webhook payloads are accepted at: {Bot api}/mappings/webhooks/{event}

### CODE REVIEW

**Comment by ghost** [comment]
Dashboard pull request: https://github.com/CreatorCoinTools/rally-discordbot-dashboard/pull/6

File: rallyrolebot/cogs/defaults_cog.py [hunk 1]
```diff
@@ -37,6 +35,115 @@ async def cog_command_error(self, ctx, error):
             type(error), error, error.__traceback__, file=sys.stderr
         )
 
+    @staticmethod
+    async def update_setting(ctx, alert, alert_nr, value, setting):
+        # check if any value is missing
+        if not alert or not alert_nr or not value:
```
Comment (Ju99ernaut) [inline rallyrolebot/cogs/defaults_cog.py]: This check may not be necessary as there's already an error implementation for missing arguments

File: rallyrolebot/cogs/defaults_cog.py [hunk 2]
```diff
@@ -37,6 +35,115 @@ async def cog_command_error(self, ctx, error):
...
+        instance['settings'][setting] = value
+        data.set_alerts_settings(ctx.guild.id, json.dumps(settings))
+
+        return await pretty_print(ctx, "Alert settings have been updated", title='Success', color=SUCCESS_COLOR)
+
+    @commands.command(
+        name='setmin',
+        help='<alert> <alert nr> <value> - Set the minimum amount for an alert'
+    )
+    async def setmin(self, ctx, alert=None, alert_nr=None, value=None):
```
Comment (Ju99ernaut) [inline rallyrolebot/cogs/defaults_cog.py]: Avoid setting `None` default for required arguments

File: rallyrolebot/cogs/defaults_cog.py [hunk 3]
```diff
@@ -37,6 +35,115 @@ async def cog_command_error(self, ctx, error):
...
+        return await self.update_setting(ctx, alert, alert_nr, value, 'minamount')
+
+    @commands.command(
+        name='setmax',
+        help='<alert> <alert nr> <value> - Set the minimum amount for an alert'
+    )
+    async def setmax(self, ctx, alert=None, alert_nr=None, value=None):
```
Comment (Ju99ernaut) [inline rallyrolebot/cogs/defaults_cog.py]: Avoid setting `None` default for required arguments

File: rallyrolebot/cogs/defaults_cog.py [hunk 4]
```diff
@@ -37,6 +35,115 @@ async def cog_command_error(self, ctx, error):
...
+        return await self.update_setting(ctx, alert, alert_nr, value, 'maxamount')
+
+    @commands.command(
+        name='settimezone',
+        help='<alert nr> <value (-12 - +12)> - Set timezone setting for daily stats message'
+    )
+    async def settimezone(self, ctx, alert_nr=None, value=None):
```
Comment (Ju99ernaut) [inline rallyrolebot/cogs/defaults_cog.py]: Avoid setting `None` default for required arguments

File: rallyrolebot/cogs/defaults_cog.py [hunk 5]
```diff
@@ -37,6 +35,115 @@ async def cog_command_error(self, ctx, error):
...
+        if not alert_nr or not value:
```
Comment (Ju99ernaut) [inline rallyrolebot/cogs/defaults_cog.py]: No need for this check as an error is already raised for missing arguments

File: rallyrolebot/cogs/defaults_cog.py [hunk 6]
```diff
@@ -37,6 +35,115 @@ async def cog_command_error(self, ctx, error):
...
+            return await pretty_print(ctx, "Missing <alert_nr> or <value>", title='Error', color=ERROR_COLOR)
+
+        return await self.update_setting(ctx, 'daily_stats', alert_nr, value, 'timezone')
+
+    @commands.command(
+        name='allcoinstats',
+        help='<day/week> - list the following stats in the coin alerts channel based on the time given'
+    )
+    async def allcoinstats(self, ctx, timeframe=''):
```
Comment (Ju99ernaut) [inline rallyrolebot/cogs/defaults_cog.py]: Add `Timeframe` convertor in `utils.convertors`

File: rallyrolebot/cogs/defaults_cog.py [hunk 7]
```diff
@@ -37,6 +35,115 @@ async def cog_command_error(self, ctx, error):
...
+        data.delete_week_old()
+        timeframe = timeframe.lower()
+        if timeframe not in ['day', 'week']:
```
Comment (Ju99ernaut) [inline rallyrolebot/cogs/defaults_cog.py]: Move this check to `utils.convetors`

**Review by Ju99ernaut — CHANGES_REQUESTED** [review]

File: rallyrolebot/cogs/defaults_cog.py [hunk 8]
```diff
@@ -37,6 +35,115 @@ async def cog_command_error(self, ctx, error):
...
+        else:
+            coin_stats = update_cog.get_week_stats(default_coin)
+
+        rewards = rally_api.get_coin_rewards(default_coin)
+        coin_image_url = rally_api.get_coin_image_url(default_coin)
+
+        extra_str = 'Today' if timeframe == 'day' else 'This Week'
+        reward_str = 'last24HourEarned' if timeframe == 'day' else 'weeklyAccumulatedReward'
+        message = {
+            "description": f"```xl\n"
```
Comment (Ju99ernaut) [inline rallyrolebot/cogs/defaults_cog.py:121]: Maybe omit `# of`

File: rallyrolebot/cogs/update_cog.py [hunk 9]
```diff
@@ -118,106 +311,215 @@ async def force_update(bot, ctx):
...
+            coin_day_stats = get_day_stats(default_coin)
+            total_stats = rally_api.get_coin_summary(default_coin)
+            rewards = rally_api.get_coin_rewards(default_coin)
+            
+            # create stats message
+            coin_image_url = rally_api.get_coin_image_url(default_coin)
+            message = {
+                "embeds": [
+                    {
+                        "description": f"```xl\n- Total # of coins: {total_stats['totalCoins']}\n\n"
```
Comment (Ju99ernaut) [inline rallyrolebot/cogs/update_cog.py]: Maybe omit `# of`

**Review by Ju99ernaut — CHANGES_REQUESTED** [review]

**Review by Ju99ernaut — CHANGES_REQUESTED** [review]
Round of decimal values, maybe 2 to 3 d.p

**Review by Ju99ernaut — APPROVED** [review]

**Comment by voanhcung** [comment]
Ok

**Review by voanhcung — COMMENTED** [review]
Ok

**Comment by voanhcung** [comment]
0xB98ceE834EF70C4F67CD5a7dCdBa89ed4948e0c5

**Pull request merged**